include/wx/textctrl.h

```diff
@@ -270,6 +270,18 @@ enum wxTextAttrLineSpacing
     wxTEXT_ATTR_LINE_SPACING_TWICE          = 20
 };
 
+#if wxUSE_SPELLCHECK
+/*!
+ * Spellcheck status values
+ */
+enum wxTextSpellcheckStatus
```

Comment:
Actually, I think this could be useful, for example on wxMSW where spell checking isn't available on earlier versions of Windows.

Reply:
The question is, how would you handle this `NOT_AVAILABLE` return value? AFAICS in exactly the same way as `ERROR`, i.e. basically just ignore it.

IOW I don't see any difference between failing to enable spell-checking because the application is running under Windows 7 or because it is running under some platform without support for it. Do you?

Reply:
Yes, you can present the user with a meaningful message that their system does not support spell checking and they should upgrade to Windows 8 or 10 if they want such functionality. 

To me, any other error implies that there is something wrong (bug / defect) in the code that should be reported to the software provider.

Not only that, if we discover other meaningful errors that can be returned in future, then we can implement them without breaking (well written) existing code. If the only option is ERROR or OKAY then we are rather in a corner.

Personally, I think the worst kind of error messages are: a) none - just silent failures where nobody knows why something didn't work; or b) generic "Something has gone wrong - contact your system administrator".

I am not sure we can solve the world's software error problems by a simple enum, but it is a step in the right direction :))

Reply:
I strongly disagree that you should ever show such a message, it's extremely annoying and not useful at all. If you really, really want to do it, I'd recommend using an approach similar to `wxWindow::IsTransparentBackgroundSupported()`, i.e. provide `IsSpellCheckingAvailable()` method that would also actually return the message explaining why it isn't.

But to me this seems like an overkill because, unlike in the transparent background case, there is nothing useful we can tell the user anyhow. But at the very least this approach would allow to possibly say something useful in the future, even if it's not useful now whereas saying "not available" without any more information is not better than saying "error" at all.

Reply:
Sorry to have written something that you strongly disagree with!! Which message though - that the user's system does not support spell checking?

I don't want to labour this point, so I will go with IsSpellCheckAvailable(). The return value can either be a bool or an enum as you wish.

However, if you disagree with this too, then I am happy to abandon it because consistency throughout the library is of overriding importance.

Reply:
There is absolutely nothing wrong with writing something I strongly disagree with! Of course, I also think that there is nothing wrong with writing that I do disagree with it, when it's the case... To avoid any ambiguity, I'm disagreeing with showing a message "Spell checking is not available on your system, upgrade to Windows 10 to enable it" to the user. I can elaborate why if you're really curious, but to me this seems such an obviously user-unfriendly thing to do, that I'm not even sure if it's needed.

Adding `IsSpellCheckingSupported()` doesn't elicit any such strong feelings on my part, I just don't think it's really necessary or going to be used much in practice, but it could be potentially useful, so please feel free to add it if you think it's worth it. However I think it should return just a bool, both for consistency and because it's hopeless to try to encode all possible failure reasons in an enum, they should be explained in a user-readable way in the output `wxString* reason` parameter, as with `IsTransparentBackgroundSupported()`.

Reply:
Sorry, I should have looked at `IsTransparentBackgroundSupported()` first. So to conclude, I will add a `bool IsSpellCheckSupported(wxString* reason = NULL)` method and return a bool from `EnableSpellCheck()`.

*Yes, quite curious, I can think of reasons to avoid such messages on an embedded system (say car GPS) but not on a desktop platform. Furthermore, the application programmer is quite at liberty to ignore the return value - I am not proposing at all that the message should be generated by wxWidgets.

Reply:
It seems obvious to me that this message is never going to convince anybody to upgrade their OS: surely if they haven't done it yet, they have some reasons for not doing it (doesn't matter whether they're good or bad) and are not going to be swayed by some nagging from your program? So what's the point of giving it, except to annoy the user? I'm really puzzled why do you think this can ever be helpful...

Reply:
IMHO in this type of situation it is better to notify someone why something they expected to work hasn't worked rather than let it silently fail. At least they know the reason, even if they do nothing about it.

A "Do not show this message again" checkbox might be appropriate in this scenario.

I guess the alternative is to Google "AmazingApp spell check is not working" and hope someone before you has had the same problem (cough, cough - Microsoft Word).

My now rather philosophical point is that by not providing any such information to the application engineer, they are forced to write non-naggy apps even if they prefer to write naggy ones. Is that good for a general purpose library? (Just to be clear, the apps can be naggy, but the library definitely shouldn't be).

It's an interesting point though. Which is better "Hmm, I wonder why that didn't work?" vs. "Computer says no."? Thank you for taking the time to discuss it.

Reply:
In the end, I am going to yield on this! If EnableSpellCheck() returns true, then it can be assumed all is fine. If it returns false, then it is because there is a bug in the application code (e.g. the control is not multiline) or the platform isn't supported.